pkg/image/registryclient/alternativesources.go

```diff
@@ -0,0 +1,229 @@
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+// PreferredImageSource chooses appropriate DockerImageReference for a Context from possible image sources gathered from
+// ImageContentSourcePolicy objects and user-passed image. PreferredImageSource will implicitly lookup from cluster and from user-given image
+// if alwaysTryAlternativeSources is true, otherwise will lookup from file if file given, from cluster if LookupClusterICSP=true, or from
+// original image-reference of user-given image (that may be different from user-given in case of mirrored images) if neither ICSP file given or lookupCluster true.
+func (c *Context) PreferredImageSource(image imagereference.DockerImageReference, alwaysTryAlternativeSources, insecure bool) (imagereference.DockerImageReference, distribution.Repository, distribution.ManifestService, error) {
+	var (
+		repo          distribution.Repository
+		replacedImage imagereference.DockerImageReference
+		manifests     distribution.ManifestService
+		err           error
+	)
+	ctx := context.TODO()
+
+	// alwaysTryAlternativeSources is false when don't want implicit alternative image source lookup
+	// LookupClusterICSP, ICSPFile indicate explicit alternative image source lookup
+	if len(c.ImageSources.ICSPFile) == 0 && !c.ImageSources.LookupClusterICSP && !alwaysTryAlternativeSources {
+		if repo, err = c.Repository(ctx, image.DockerClientDefaults().RegistryURL(), image.RepositoryName(), insecure); err != nil {
+			return imagereference.DockerImageReference{}, nil, nil, err
+		}
+		manifests, err = repo.Manifests(ctx)
+		return image, repo, manifests, err
+	}
+
+	var altSources []imagereference.DockerImageReference
+	// error if given an icsp file and don't successfully connect to image repository
+	if len(c.ImageSources.ICSPFile) > 0 {
+		if err = c.AddImageSourcePoliciesFromFile(); err != nil {
+			return imagereference.DockerImageReference{}, nil, nil, err
+		}
+		altSources, err = c.AddImageSources(image)
+		if err != nil {
+			return imagereference.DockerImageReference{}, nil, nil, err
+		}
+		for _, icsRef := range altSources {
+			replacedImage = replaceImage(icsRef, image)
+			// if not successful, error will be handled below
+			if repo, err = c.Repository(ctx, replacedImage.DockerClientDefaults().RegistryURL(), replacedImage.RepositoryName(), insecure); err == nil {
+				manifests, err = repo.Manifests(ctx)
+				return replacedImage, repo, manifests, err
+			}
+		}
+	}
+
+	// error if explicitly asked to look in cluster (from command line flag) and didn't connect to image repository
+	if c.ImageSources.LookupClusterICSP {
+		// now try to look for ICSPs from cluster
+		if c.ImageSources.ICSPClientFn == nil {
```

Comment:
I think you could get away with just passing `ICSPClientFn` which would be a sign that a user requested that, wdyt?
This will mean that in oc in `SecurityOptions#Complete` you'd initialize `o.ICSPClientFn` only if `--lookup-cluster-icsp` was requested. This will simplify this piece of code which is the only relying on the presence of `LookupClisterICSP`.

Reply:
yes, I had to rework use of flags in oc a bit, but now we don't need the LookupClusterICSP.